src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java

```diff
@@ -781,6 +786,157 @@ private void mergeStagingTable(TableId stagingTableId, TableBlob blob,
+       * WHEN MATCHED AND D.op IN (“INSERT”, “UPDATE”)
+       *              AND D._sequence_num > T._sequence_num
+       *    UPDATE _sequence_num = D._sequence_num, id = D.id, name = D.name
+       * WHEN NOT MATCHED
+       *    INSERT (_sequence_num, _row_id, _is_deleted, id, name) VALUES
+       *           (D._sequence_num, D._row_id, false, D.id, D.name)
+       *
+       * Following are the differences compared to merge query for ORDERED events.
+       * 1. join condition used for the merge query includes _row_id.
+       * 2. when match happens for DELETE operation, update _is_deleted in the target table to true
+       * 3. when no match insert the new row
+       */
+      mergeCondition = " T._row_id = D._row_id ";
+      deleteOperation = "  UPDATE SET _is_deleted = true ";
+    } else {
+      /*
+       * Merge query will be of the following form:
+       *
+       * MERGE [target table] as T
+       * USING ($DIFF_QUERY) as D
+       * ON T.id = D._before_id
+       * WHEN MATCHED AND D._op = "DELETE" THEN
+       *   DELETE
+       * WHEN MATCHED AND D._op IN ("INSERT", "UPDATE") THEN
+       *   UPDATE id = D.id, name = D.name
+       * WHEN NOT MATCHED AND D._op IN ("INSERT", "UPDATE") THEN
+       *   INSERT (id, name) VALUES (id, name)
+       *
+       * Following are the differences compared to merge query for UN-ORDERED events.
+       * 1. join condition used for the merge query includes primary keys.
+       * 2. when match happens for DELETE operation, issue DELETE command
+       */
+      mergeCondition = primaryKeys.stream()
+        .map(name -> String.format("T.%s = D._before_%s", name, name))
+        .collect(Collectors.joining(" AND "));
+
+      deleteOperation = "  DELETE";
+    }
+
+    // target table schema always contains nullable fields such as _is_deleted, _row_id.
+    // however for ORDERED source, these fields are not in the staging table, so filter out these fields before
+    // we perform INSERT/UPDATE queries based on the values from the staging table.
+    final Predicate<Schema.Field> predicate = field -> {
+      // filter out _is_deleted field for operations INSERT and UPDATE as it will be set when
+      // the operation is DELETE.
+      if (field.getName().equals(Constants.IS_DELETED)) {
+        return false;
+      }
+      if (sourceEventOrdering != SourceProperties.Ordering.ORDERED) {
+        return true;
+      }
+      // filter out the _row_id field for ORDERED source, as this field will not be present in the staging table.
+      return !field.getName().equals(Constants.ROW_ID);
+    };
+
+    return "MERGE " +
+      targetTableId.getDataset() + "." + targetTableId.getTable() + " as T\n" +
+      "USING (" + diffQuery + ") as D\n" +
+      "ON " + mergeCondition + "\n" +
+      "WHEN MATCHED AND D._op = \"DELETE\" THEN\n" +
+      deleteOperation + "\n" +
+      "WHEN MATCHED AND D._op IN (\"INSERT\", \"UPDATE\") AND D._sequence_num > T._sequence_num THEN\n" +
```

Comment:
"INSERT" looks not necessary here ?

Reply:
Its possible in the scenario where in the same batch if the record with primary key x is deleted and new record with the same primary key is inserted. New record will have _op as INSERT.

Reply:
the diff query will have two row one with _op=delete and one with _op=insert in this case. only the row with _op = delete has value in "_before" columns and the row with _op=insert won't have any value in "_before" columns. Thus the row  with _op=delete will match the existing row in target table and that row got deleted. row with _op=insert won't match any existing row in target table.

Reply:
ah. I thought comment is for `NOT MATCHED` where we do need insert and update. yes here we dont. good catch! fixed the code.